titan/features/partner_tracing.py

```diff
@@ -21,35 +62,167 @@ def update_agent(self, model: "model.TITAN"):
         """
+        
         params = self.agent.location.params.partner_tracing
-
+        agent_exposure = getattr(self.agent, params.exposure)
+        
+        # Starting Jan 2021, use improved PS parameters
+        if model.time >= 48:
+            params[self.agent.race].participate_prob = params[self.agent.race].participate_prob_improved
+            params.tracing_prob = params.tracing_prob_improved
+            params[self.agent.race].contact_prob = params[self.agent.race].contact_prob_improved
+            params[self.agent.race].partner_participate_prob = params[self.agent.race].partner_participate_prob_improved
+            params.treatment_prob = params.treatment_prob_improved
+            params[self.agent.race].prep_prob = params[self.agent.race].prep_prob_improved
+            params.re_haart_prob = params.re_haart_prob_improved
+            
         if model.time < params.start_time or model.time > params.stop_time:
             return
 
-        agent_exposure = getattr(self.agent, params.exposure)
+        # if the agent was diagnosed with hiv in the previous time step and is willing to participate, trace their partners
+        if (
+            agent_exposure.dx
+            and agent_exposure.dx_time == model.time - 1
+        ):  
+            # step 1 
+            self.participate = True if model.run_random.random() < params[self.agent.race].participate_prob else False
 
-        if agent_exposure.active:
-            # was this agent diagnosed with the target exposure last time step
-            if agent_exposure.dx and agent_exposure.dx_time == model.time - 1:
-                for ptnr in self.agent.get_partners(params.bond_type):
-                    partner_exposure = getattr(ptnr, params.exposure)
-                    if (
-                        not partner_exposure.dx
-                        and model.run_random.random() < params.prob
-                    ):
-                        ptnr.partner_tracing.active = True  # type: ignore[attr-defined]
-                        ptnr.partner_tracing.time = model.time  # type: ignore[attr-defined]
-
-            # second chance at diagnosis if traced
+            if self.participate:
+                self.participate_time = model.time
+                
+                self.trace_partners(params, model)
+                
+                        
+        # if this agent is traced (i.e. has been contacted by PS) and comes back for testing
+        if (
+            self.active
+            and self.time < model.time
+            and model.run_random.random() < params[self.agent.race].partner_participate_prob
+        ):  
+            # mark agents who participate
+            self.ps_participant = True
+            self.ps_participant_time = model.time
+            
+            # if this traced agent was already previously diagnosed, mark as previously diagnosed (every agent is allowed to enter this if statemnet no more than once)
```

Comment:
What do you mean by "every agent is allowed to enter this if statemnet no more than once"?